File: clients/file-system/py/src/file_system/validation/paths.py
"""Validate raw inputs before filesystem operations."""

from __future__ import annotations

from pathlib import Path

from pydantic import ValidationError

from file_system.core.error_format import format_validation_detail
from file_system.core.errors import AppError, ErrorCodes
from file_system.core.results import Failure, Success
from file_system.core.types import FsResult
from file_system.domain.enums import TextEncoding
from file_system.domain.models import BytesWriteRequest, PathRequest, TextWriteRequest


def validate_path(path: str | Path) -> FsResult[PathRequest]:
    """Validate a filesystem path at the trust boundary."""
    try:
        return Success(PathRequest(path=Path(path)))
    except ValidationError as exc:
        return Failure(
            AppError(
                code=ErrorCodes.VALIDATION,
                message="Invalid path.",
                detail=format_validation_detail(exc),
            ),
        )


def validate_text_write(
    path: str | Path,
    text: str,
    encoding: TextEncoding,
) -> FsResult[TextWriteRequest]:
    """Validate text write inputs."""
    try:
        return Success(
            TextWriteRequest(path=Path(path), text=text, encoding=encoding),
        )
    except ValidationError as exc:
        return Failure(
            AppError(
                code=ErrorCodes.VALIDATION,
                message="Invalid text write request.",
                detail=format_validation_detail(exc),
            ),
        )


def validate_bytes_write(path: str | Path, data: bytes) -> FsResult[BytesWriteRequest]:
    """Validate binary write inputs."""
    try:
        return Success(BytesWriteRequest(path=Path(path), data=data))
    except ValidationError as exc:
        return Failure(
            AppError(
                code=ErrorCodes.VALIDATION,
                message="Invalid binary write request.",
                detail=format_validation_detail(exc),
            ),
        )

File: clients/file-system/ts/src/validation/paths.ts
import { err, ok } from "neverthrow";

import { formatValidationDetail } from "../core/error-format.js";
import { ErrorCodes, appError } from "../core/errors.js";
import type { FsResult } from "../core/types.js";
import type { TextEncoding } from "../domain/enums.js";
import {
  BytesWriteRequestSchema,
  PathRequestSchema,
  TextWriteRequestSchema,
  type BytesWriteRequest,
  type PathRequest,
  type TextWriteRequest,
} from "../domain/schemas.js";

export function validatePath(path: string): FsResult<PathRequest> {
  const parsed = PathRequestSchema.safeParse({ path });
  if (!parsed.success) {
    return err(
      appError(
        ErrorCodes.VALIDATION,
        "Invalid path.",
        formatValidationDetail(parsed.error),
      ),
    );
  }
  return ok(parsed.data);
}

export function validateTextWrite(
  path: string,
  text: string,
  encoding: TextEncoding,
): FsResult<TextWriteRequest> {
  const parsed = TextWriteRequestSchema.safeParse({ path, text, encoding });
  if (!parsed.success) {
    return err(
      appError(
        ErrorCodes.VALIDATION,
        "Invalid text write request.",
        formatValidationDetail(parsed.error),
      ),
    );
  }
  return ok(parsed.data);
}

export function validateBytesWrite(
  path: string,
  data: Buffer,
): FsResult<BytesWriteRequest> {
  const parsed = BytesWriteRequestSchema.safeParse({ path, data });
  if (!parsed.success) {
    return err(
      appError(
        ErrorCodes.VALIDATION,
        "Invalid binary write request.",
        formatValidationDetail(parsed.error),
      ),
    );
  }
  return ok(parsed.data);
}
